primitives/arithmetic/src/fixed.rs

```diff
@@ -0,0 +1,1308 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use sp_std::{ops::{self, Add, Sub, Mul, Div}, fmt::Debug, prelude::*, convert::{TryInto, TryFrom}};
+use codec::{Encode, Decode};
+use crate::{
+	helpers_128bit::multiply_by_rational, PerThing,
+	traits::{
+		SaturatedConversion, CheckedSub, CheckedAdd, CheckedMul, CheckedDiv, CheckedNeg,
+		Bounded, Saturating, UniqueSaturatedInto, Zero, One
+	},
+};
+
+#[cfg(feature = "std")]
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+
+pub trait FixedPointOperand: Copy + Clone + Bounded + Zero + Saturating
+	+ PartialOrd + UniqueSaturatedInto<u128> + TryFrom<u128> + CheckedNeg {}
+
+impl FixedPointOperand for i128 {}
+impl FixedPointOperand for u128 {}
+impl FixedPointOperand for i64 {}
+impl FixedPointOperand for u64 {}
+impl FixedPointOperand for i32 {}
+impl FixedPointOperand for u32 {}
+impl FixedPointOperand for i16 {}
+impl FixedPointOperand for u16 {}
+impl FixedPointOperand for i8 {}
+impl FixedPointOperand for u8 {}
+
+/// Something that implements a fixed point with an arbitrary precision `DIV`,
+/// meaning that `1 / DIV` can be represented. `DIV` should be a power of `10`.
+pub trait FixedPointNumber:
+	Sized + Copy + Default + Debug
+	+ Saturating + Bounded
+	+ Eq + PartialEq + Ord + PartialOrd
+	+ CheckedSub + CheckedAdd + CheckedMul + CheckedDiv
+	+ Add + Sub + Div + Mul
+{
+	/// The underlying data type used for this fixed point number.
+	type Inner: Copy + Debug + Zero + One + PartialOrd;
```

Comment:
```suggestion
	type Inner: Copy + Debug + Zero + One + PartialOrd + Saturating;
```

will help move impls to trait.

Reply:
yeah, I was leaving it for last